src/gurobi_optimods/line_optimization.py

```diff
@@ -0,0 +1,512 @@
+                foundV = False
+                for u, v in linepaths[l]:
+                    if s == u:
+                        for f in frequencies:
+                            capFS += x[l, f] * lines[l]["capacity"] * f
+                        if foundU == False:
+                            for f in frequencies:
+                                linesFS += x[l, f]
+                            foundU = True
+                    if s == v:
+                        for f in frequencies:
+                            capTS += x[l, f] * lines[l]["capacity"] * f
+                        if foundV == False:
+                            for f in frequencies:
+                                linesTS += x[l, f]
+                            foundV = True
+            if demandFromS > 0:
+                model.addConstr(capFS >= demandFromS, name="capFrom" + str(s))
+                model.addConstr(linesFS >= 1, name="linesFrom" + str(s))
+            if demandToS > 0:
+                model.addConstr(capTS >= demandToS, name="capTo" + str(s))
+                model.addConstr(linesTS >= 1, name="linesTo" + str(s))
+
+        model.optimize()
+
+        # prepare return values
+        objCost = -1
+        linesOut = []
+        if model.Status == gp.GRB.OPTIMAL:
+            objCost = model.objVal
+            # model.setParam(gp.GRB.Param.ObjNumber, 1)
+            # objTime = model.objVal
+            # save optimal line plan (after both objectives) in list of tuples
+            # print non-zero solution values of line variables
+            for i in linepaths.index:
+                for j in frequencies:
+                    if x[i, j].X > 0.5:
+                        linesOut.append((i, j))
+        return objCost, linesOut
+
+
+def plot_lineplan(
+    node_data: pd.DataFrame,
+    edge_data: pd.DataFrame,
+    linepath_data: pd.DataFrame,
+    line_plan: list,
+):
+    """Visualize a line plan with at most 20 lines. The figure is opened in a browser
+
+    Parameters
+    ----------
+    node_data : DataFrame
+        DataFrame with information on the nodes/stations. The frame must include "source".
+    edge_data : DataFrame
+        DataFrame with edges / connections between stations and associated attributes.
+        The frame must include "source", "target", and "time"
+    linepath_data : DataFrame
+        DataFrame with information on the line routes/paths.
+        It must include "linename", "edgeSource", and "edgeTarget".
+    line_plan: List
+        A solution of the line optimization, i.e., a list with linenames and associated frequencies.
+
+    """
+    if mpl is None or nx is None:
+        logger.info(
+            "Plot not possible: networkx, matplotlib and matplotlib.pyplot are required for plotting the line plan"
+        )
+    if len(line_plan) >= 20:
+        logger.info(
+            "Line plan has more than 20 lines, only visualize line plans with at most 20 lines"
+        )
+        return
```

Comment:
In both these cases, you should raise an exception instead of writing to the log. `raise RuntimeError(message)` is probably the way to go here.

In general, you don't want code to fail silently, even if you only expect a user to use this function in interactive mode.

Reply:
Could you also add a comment here explaining the 20 line limitation? (I guess it's just due to limited availability of colours? Or beyond 20 lines does it just look like complete chaos?)